filters/auth/oauth.go

```diff
@@ -0,0 +1,429 @@
+package auth
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+	logfilter "github.com/zalando/skipper/filters/log"
+	"golang.org/x/oauth2"
+)
+
+type roleCheckType int
+
+const (
+	checkAnyScopes roleCheckType = iota
+	checkAllScopes
+	checkAnyKV
+	checkAllKV
+	checkUnknown
+)
+
+type rejectReason string
+
+const (
+	missingBearerToken rejectReason = "missing-bearer-token"
+	authServiceAccess  rejectReason = "auth-service-access"
+	invalidToken       rejectReason = "invalid-token"
+	invalidRealm       rejectReason = "invalid-realm"
+	invalidScope       rejectReason = "invalid-scope"
+)
+
+// TODO: discuss these names, because these are the filter names used by the endusers
+const (
+	AuthAnyScopeName = "authAnyScope"
+	AuthAllScopeName = "authAllScope"
+	AuthAnyKVName    = "authAnyKV"
+	AuthAllKVName    = "authAllKV"
+	AuthUnknown      = "authUnknown"
+
+	authHeaderName = "Authorization"
+	realmKey       = "realm"
+	scopeKey       = "scope"
+	uidKey         = "uid"
+)
+
+type (
+	authClient struct{ urlBase string }
+
+	spec struct {
+		typ        roleCheckType
+		cfg        oauth2.Config
+		authClient *authClient
+	}
+
+	filter struct {
+		typ        roleCheckType
+		authClient *authClient
+		realm      string
+		scopes     []string
+		kv         kv
+	}
+	kv map[string]string
+)
+
+var (
+	errInvalidAuthorizationHeader = errors.New("invalid authorization header")
+	errInvalidToken               = errors.New("invalid token")
+)
+
+func getToken(r *http.Request) (string, error) {
+	const b = "Bearer "
+	h := r.Header.Get(authHeaderName)
+	if !strings.HasPrefix(h, b) {
+		return "", errInvalidAuthorizationHeader
+	}
+
+	return h[len(b):], nil
+}
+
+func unauthorized(ctx filters.FilterContext, uname string, reason rejectReason) {
+	ctx.StateBag()[logfilter.AuthUserKey] = uname
+	ctx.StateBag()[logfilter.AuthRejectReasonKey] = string(reason)
+	ctx.Serve(&http.Response{StatusCode: http.StatusUnauthorized})
+}
+
+func authorized(ctx filters.FilterContext, uname string) {
+	ctx.StateBag()[logfilter.AuthUserKey] = uname
+}
+
+func getStrings(args []interface{}) ([]string, error) {
+	s := make([]string, len(args))
+	var ok bool
+	for i, a := range args {
+		s[i], ok = a.(string)
+		if !ok {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	return s, nil
+}
+
+// all checks that all strings in the left are also in the
+// right. Right can be a superset of left.
+func all(left, right []string) bool {
+	if len(left) > len(right) {
+		return false
+	}
+
+	for _, l := range left {
+		for _, r := range right {
+			if l != r {
+				return false
```

Comment:
this will return false for any larger r than len=1

Reply:
fixed